File: yt/yt/flow/library/python/companion/state.py
"""State classes: State, ExternalState, StatesHolder."""

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, Generic, Optional, TypeVar

from .row import Payload, TableSchema
from .wire_protocol import UnversionedRow


@dataclass
class State:
    """Wrapper for binary internal state.

    Next to the wire bytes the entry keeps the decoded value once an accessor hands it out, and
    the encoder that turns it back into bytes, so that the changes made to that value in place
    are picked up by :meth:`sync_bytes`.
    """

    reset: bool = False
    state: Optional[bytes] = None
    value: Any = field(default=None, compare=False, repr=False)
    codec: Any = field(default=None, compare=False, repr=False)
    encode: Optional[Callable[[Any], bytes]] = field(default=None, compare=False, repr=False)

    def get_value(self, codec: Any, decode: Callable[[bytes], Any]) -> Any:
        """Value decoded from the wire bytes, memoized per |codec|: every accessor of that codec
        gets the same object, while an accessor of another one decodes the bytes anew.

        Decoding anew takes the bytes as they arrived, so it drops a change made in place through
        the previous codec.
        """
        if self.value is None or self.codec != codec:
            self.value = decode(self.state)
            self.codec = codec
            self.encode = None
        return self.value

    def get_mutable_value(self, codec: Any, decode: Callable[[bytes], Any], encode: Callable[[Any], bytes]) -> Any:
        """Value like :meth:`get_value`, tracked for changes: |encode| is kept, so that
        :meth:`sync_bytes` re-encodes the value and the changes made to it in place reach the wire.
        """
        value = self.get_value(codec, decode)
        self.encode = encode
        return value

    def sync_bytes(self) -> bool:
        """Re-encode a tracked value and report whether the wire bytes changed."""
        if self.encode is None:
            return False
        encoded = self.encode(self.value)
        if encoded == self.state:
            return False
        self.state = encoded
        return True


@dataclass
class ExternalState:
    """Wrapper for external state backed by Payload."""

    reset: bool = False
    state: Optional[Payload] = None

    def sync_bytes(self) -> bool:
        """External state is written through the accessor's ``set()``, never tracked in place."""
        return False


# Sentinel for reset external state.
EXTERNAL_STATE_RESET = ExternalState(reset=True, state=None)


T = TypeVar("T", State, ExternalState)


class StatesHolder(Generic[T]):
    """Holder of raw binary state representation, keyed by UnversionedRow."""

    def __init__(
        self,
        name: str,
        key_schema: Optional[TableSchema] = None,
        state_schema: Optional[TableSchema] = None,
    ):
        self.name = name
        self.key_schema = key_schema
        self.state_schema = state_schema
        self._states: Dict[tuple, tuple[UnversionedRow, T]] = {}
        # Entries changed during the current epoch: via set() (writes from state accessors), or
        # in place through a mutable value, as found by collect_modified().
        # Keys populated from the incoming request via load() are intentionally excluded so that
        # only modified states are sent back in the response.  The entry is kept, not just its
        # key: a state attached over a pending reset must not turn that reset into a write.
        self._modified: Dict[tuple, tuple[UnversionedRow, T]] = {}

    def _row_key(self, row: UnversionedRow) -> tuple:
        """Create a hashable key from an UnversionedRow."""
        parts = []
        for v in row.values:
            parts.append((v.column_id, v.type, v.value if not isinstance(v.value, bytearray) else bytes(v.value)))
        return tuple(parts)

    def set(self, key: UnversionedRow, value: T):
        """Set a value for key and mark it modified (so it is sent back in the response)."""
        row_key = self._row_key(key)
        self._states[row_key] = (key, value)
        self._modified[row_key] = (key, value)

    def load(self, key: UnversionedRow, value: T):
        """Load a value from the request WITHOUT marking it modified.

        Used to populate the holder from the incoming request: such states must not be echoed
        back unless a state accessor changes them.
        """
        self._states[self._row_key(key)] = (key, value)

    def get(self, key: UnversionedRow) -> Optional[T]:
        entry = self._states.get(self._row_key(key))
        return entry[1] if entry else None

    def items(self):
        """Iterate over (key, value) pairs for all states."""
        for row, val in self._states.values():
            yield row, val

    def modified_items(self):
        """Iterate over (key, value) pairs for states modified this epoch."""
        for row, value in self._modified.values():
            yield row, value

    def collect_modified(self) -> bool:
        """Re-encode the values handed out as mutable and report whether the holder has anything
        to send back: a value changed in place counts as modified, one that re-encodes to the
        bytes it arrived with does not.
        """
        for row_key, entry in self._states.items():
            if entry[1].sync_bytes():
                self._modified[row_key] = entry
        return self.has_modified()

    def has_modified(self) -> bool:
        """Whether any state was modified during the current epoch; call collect_modified() first
        to pick up the values changed in place.
        """
        return bool(self._modified)

File: yt/go/flow/state.go
package flow

import (
	"iter"

	"go.ytsaurus.tech/library/go/core/xerrors"
	"go.ytsaurus.tech/yt/go/wire"
)

var (
	// ErrNoStateSchema reports an external state declared without the schema of its rows.
	ErrNoStateSchema = xerrors.NewSentinel("external state has no schema")

	// ErrStateNotRead reports an external state absent from the request.
	ErrStateNotRead = xerrors.NewSentinel("external state was not read by this request")
)

// StateValue is a value stored by StatesHolder.
type StateValue[T any] interface {
	Cleared() T
}

// InternalState is an entry of a keyed internal state.
type InternalState struct {
	// Reset asks the worker to drop the state stored for the key; Data is then ignored.
	Reset bool
	Data  []byte
}

// Cleared returns the entry that deletes the state stored for a key.
func (s InternalState) Cleared() InternalState {
	return InternalState{Reset: true}
}

// ExternalState is an entry of a keyed external state.
type ExternalState struct {
	// Reset asks the worker to delete the row stored for the key; Value is then ignored.
	Reset bool
	Value Payload
}

// Cleared returns the entry that deletes the row stored for a key.
func (s ExternalState) Cleared() ExternalState {
	return ExternalState{Reset: true}
}

type stateEntry[T any] struct {
	key      Payload
	value    T
	modified bool
}

// StatesHolder holds one keyed state for a request.
// Modified returns only entries changed after loading.
type StatesHolder[T StateValue[T]] struct {
	name        string
	stateSchema Schema

	index   map[string]int
	entries []stateEntry[T]

	modified int
}

func newInternalStatesHolder(name string) *StatesHolder[InternalState] {
	return &StatesHolder[InternalState]{name: name, index: map[string]int{}}
}

func newExternalStatesHolder(name string, stateSchema Schema) (*StatesHolder[ExternalState], error) {
	if stateSchema.Len() == 0 {
		return nil, xerrors.Errorf("flow: %w: state %q", ErrNoStateSchema, name)
	}
	return &StatesHolder[ExternalState]{
		name:        name,
		stateSchema: stateSchema,
		index:       map[string]int{},
	}, nil
}

// Name returns the state name this holder stores entries of.
func (h *StatesHolder[T]) Name() string {
	return h.name
}

// StateSchema returns the external state row schema.
func (h *StatesHolder[T]) StateSchema() Schema {
	return h.stateSchema
}

// Len returns the number of entries.
func (h *StatesHolder[T]) Len() int {
	return len(h.entries)
}

// Get returns the entry stored for key.
func (h *StatesHolder[T]) Get(key Payload) (T, bool) {
	var zero T
	rowKey, err := encodeStateKey(key)
	if err != nil {
		return zero, false
	}
	i, ok := h.index[rowKey]
	if !ok {
		return zero, false
	}
	return h.entries[i].value, true
}

// Load stores an entry without marking it modified.
func (h *StatesHolder[T]) Load(key Payload, value T) error {
	return h.put(key, value, false)
}

// Set stores a user write for key and marks it modified.
func (h *StatesHolder[T]) Set(key Payload, value T) error {
	return h.put(key, value, true)
}

// Clear records the deletion of the state stored for key.
func (h *StatesHolder[T]) Clear(key Payload) error {
	var zero T
	return h.Set(key, zero.Cleared())
}

// HasModified reports whether any entry was written during this request.
func (h *StatesHolder[T]) HasModified() bool {
	return h.modified > 0
}

// All iterates over every entry held, in the order the keys were first seen.
func (h *StatesHolder[T]) All() iter.Seq2[Payload, T] {
	return func(yield func(Payload, T) bool) {
		for i := range h.entries {
			if !yield(h.entries[i].key, h.entries[i].value) {
				return
			}
		}
	}
}

// Modified iterates over modified entries in insertion order.
func (h *StatesHolder[T]) Modified() iter.Seq2[Payload, T] {
	return func(yield func(Payload, T) bool) {
		for i := range h.entries {
			if !h.entries[i].modified {
				continue
			}
			if !yield(h.entries[i].key, h.entries[i].value) {
				return
			}
		}
	}
}

func (h *StatesHolder[T]) put(key Payload, value T, modified bool) error {
	rowKey, err := encodeStateKey(key)
	if err != nil {
		return xerrors.Errorf("flow: state %q: %w", h.name, err)
	}

	if i, ok := h.index[rowKey]; ok {
		h.entries[i].value = value
		if modified && !h.entries[i].modified {
			h.entries[i].modified = true
			h.modified++
		}
		return nil
	}

	h.index[rowKey] = len(h.entries)
	h.entries = append(h.entries, stateEntry[T]{key: key, value: value, modified: modified})
	if modified {
		h.modified++
	}
	return nil
}

func encodeStateKey(key Payload) (string, error) {
	encoded, err := wire.MarshalRowProto(key.row)
	if err != nil {
		return "", xerrors.Errorf("unusable state key: %w", err)
	}
	return string(encoded), nil
}
